prometheus/echoprometheus/prometheus.go

```diff
@@ -0,0 +1,458 @@
+	// metrics that need data to be stored for AfterNext.
+	BeforeNext func(c echo.Context)
+
+	// AfterNext is callback that is executed after next middleware/handler returns. Useful for case when you have own
+	// metrics that need incremented/observed.
+	AfterNext func(c echo.Context, err error)
+
+	timeNow func() time.Time
+}
+
+type LabelValueFunc func(c echo.Context, err error) string
+
+// HandlerConfig contains the configuration for creating HTTP handler for metrics.
+type HandlerConfig struct {
+	// Gatherer sets the prometheus.Gatherer instance the middleware will use when generating the metric endpoint handler.
+	// Defaults to: prometheus.DefaultGatherer
+	Gatherer prometheus.Gatherer
+}
+
+// PushGatewayConfig contains the configuration for pushing to a Prometheus push gateway.
+type PushGatewayConfig struct {
+	// PushGatewayURL is push gateway URL in format http://domain:port
+	PushGatewayURL string
+
+	// PushInterval in ticker interval for pushing gathered metrics to the Gateway
+	// Defaults to: 1 minute
+	PushInterval time.Duration
+
+	// Gatherer sets the prometheus.Gatherer instance the middleware will use when generating the metric endpoint handler.
+	// Defaults to: prometheus.DefaultGatherer
+	Gatherer prometheus.Gatherer
+
+	// ErrorHandler is function that is called when errors occur. When callback returns error StartPushGateway also returns.
+	ErrorHandler func(err error) error
+
+	// ClientTransport specifies the mechanism by which individual HTTP POST requests are made.
+	// Defaults to: http.DefaultTransport
+	ClientTransport http.RoundTripper
+}
+
+// NewHandler creates new instance of Handler using Prometheus default registry.
+func NewHandler() echo.HandlerFunc {
+	return NewHandlerWithConfig(HandlerConfig{})
+}
+
+// NewHandlerWithConfig creates new instance of Handler using given configuration.
+func NewHandlerWithConfig(config HandlerConfig) echo.HandlerFunc {
+	if config.Gatherer == nil {
+		config.Gatherer = prometheus.DefaultGatherer
+	}
+	h := promhttp.HandlerFor(config.Gatherer, promhttp.HandlerOpts{})
+
+	if r, ok := config.Gatherer.(prometheus.Registerer); ok {
+		h = promhttp.InstrumentMetricHandler(r, h)
+	}
+
+	return func(c echo.Context) error {
+		h.ServeHTTP(c.Response(), c.Request())
+		return nil
+	}
+}
+
+// NewMiddleware creates new instance of middleware using Prometheus default registry.
+func NewMiddleware(subsystem string) echo.MiddlewareFunc {
+	return NewMiddlewareWithConfig(MiddlewareConfig{Subsystem: subsystem})
+}
+
+// NewMiddlewareWithConfig creates new instance of middleware using given configuration.
+func NewMiddlewareWithConfig(config MiddlewareConfig) echo.MiddlewareFunc {
+	mw, err := config.ToMiddleware()
+	if err != nil {
+		panic(err)
+	}
+	return mw
+}
+
+// ToMiddleware converts configuration to middleware or returns an error.
+func (conf MiddlewareConfig) ToMiddleware() (echo.MiddlewareFunc, error) {
```

Comment:
Does this have to be public? I guess most use cases are using NewMiddlewareWithConfig anyway

Reply:
This is to allowing to handle problems with middleware creation without panics.  Similar to https://github.com/labstack/echo-jwt/blob/95b0b607987a3bed484870b2052ef212763742a4/jwt.go#L173